src/main/java/com/imaee/propinq/properties/services/implementations/PropertyTypeService.java

```diff
@@ -0,0 +1,100 @@
+        Optional<PropertyType> propertyTypeOptional = propertyTypeRepository.findByName(model.getName());
+
+        if(propertyTypeOptional.isPresent()){
+            PropertyType existingPropertyType = propertyTypeOptional.get();
+            if(existingPropertyType.isState() == PropertyType.REMOVED){
+                existingPropertyType.setState(PropertyType.ACTIVE);
+                existingPropertyType.setName(model.getName());
+                existingPropertyType.setDescription(model.getDescription());
+                return PropertyTypeMapper.toPropertyTypeResponse(propertyTypeRepository.save(existingPropertyType));
+            }else {
+                throw new ResponseStatusException(HttpStatus.BAD_REQUEST, MSG_ALREADY_EXISTS);
+            }
+        }
+        return PropertyTypeMapper.toPropertyTypeResponse(propertyTypeRepository.save(model));
+    }
+
+    @Override
+    public PropertyTypeResponse updatePropertyType(PropertyTypeRequest propertyTypeRequest, UUID id) {
+        PropertyType model = PropertyTypeMapper.toPropertyType(propertyTypeRequest);
+        Optional<PropertyType> propertyTypeOptional = propertyTypeRepository.findById(id);
+
+        if (propertyTypeOptional.isPresent()) {
+            if (propertyTypeOptional.get().isState() == PropertyType.REMOVED) {
+                {
+                    throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "NO SE PUDO ACTUALIZAR, EL TIPO DE VIVIENDA ESTÁ ELIMINADO");
+                }
+            }
+            PropertyType propertyType = propertyTypeOptional.get();
+            propertyType.setName(model.getName());
+            propertyType.setDescription(model.getDescription());
+            return PropertyTypeMapper.toPropertyTypeResponse(propertyTypeRepository.save(propertyType));
+        }
+        throw new ResponseStatusException(HttpStatus.BAD_REQUEST, MSG_NOT_EXISTS);
+    }
+
+    @Override
+    public ResponseEntity<Void> deletePropertyType(UUID id) {
+        PropertyType model = propertyTypeRepository.findById(id).orElse(null);
+        if (model == null || model.isState() == PropertyType.REMOVED) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "NO SE PUDO ELIMINAR, EL TIPO DE VIVIENDA YA ESTÁ ELIMINADO");
+        }
+
+        model.setState(PropertyType.REMOVED);
+        propertyTypeRepository.save(model);
+        return ResponseEntity.noContent().build();
+    }
+
+    @Override
+    public ResponseEntity<Void> recoverPropertyType(UUID id) {
+        PropertyType model = propertyTypeRepository.findById(id).orElse(null);
+        if (model == null || model.isState() == PropertyType.ACTIVE) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "NO SE PUDO RECUPERAR, EL TIPO DE VIVIENDA YA ESTÁ ACTIVO");
+        }
+        model.setState(PropertyType.ACTIVE);
+        propertyTypeRepository.save(model);
+        return ResponseEntity.ok().build(); //respuesta con 200
+    }
```

Comment:
Tenemos un comportamiento bastante parecido al anterior

```suggestion
    @Override
    public void recoverPropertyType(UUID id) {
        final var propertyType = findByIdOrThrowException(id);
        throwExceptionIfPropertyTypeIsNotDeleted(propertyType);
        // Agreguemos una validación para ver si existe el mismo PropertyType pero activo
        propertyType.setDeleted(false);
        propertyTypeRepository.save(propertyType);
    }
    
    private void throwExceptionIfPropertyTypeIsDeleted(PropertyType propertyType) {
        if (!propertyType.isDeleted())
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, MSG_IS_NOT_DELETED);
    }
```